builder/dispatchers.go

```diff
@@ -226,10 +226,16 @@ func run(b *Builder, args []string, attributes map[string]bool, original string)
 	// set Cmd manually, this is special case only for Dockerfiles
 	b.Config.Cmd = config.Cmd
 	runconfig.Merge(b.Config, config)
+	// set build-time environment for 'run'. We let dockerfile
+	// environment override build-time environment.
+	env := b.Config.Env
```

Comment:
By "this" do you mean my proposal or this entire PR?  Because even setting env vars for RUN would seem to violate what you said.